### DISCUSSION TITLE
NetBox v3.0 and roadmap

### DISCUSSION THREAD

**Discussion by jeremystretch** [netbox-community/netbox#6500] [General]
Hey folks! I wanted to share a quick update regarding the next NetBox release and our new roadmap.

### NetBox v3.0

Those who have been following NetBox's development closely may recall that we had originally planned the UI overhaul (#5893) and secrets plugin (#5278) for the v2.11 release, but quickly realized that it would be taking on too much for a single release. So, we pushed those initiatives back to a planned v2.12 release.

Recently, however, we realized that now would be the perfect time to make the transition to NetBox v3.0. The UI overhaul and detachment of the secrets functionality obviously represent huge changes to the project. Further, since a large amount of this work has already been completed, we find some available time to commit to other major changes: namely the addition of a GraphQL interface (#2007) and ditching the Django admin UI (#6466). Taken together, we believe these more than justify the bump to v3.0.

What does this mean for end users? Honestly, not a whole lot. We don't anticipate any major breaking changes; at least nothing more significant that upgrading from v2.10 to v2.11, for example. But we wanted to ensure the version bump is communicated ahead of time so that users aren't caught off-guard. We also want to ensure that plugin authors are aware of the upcoming UI changes and can adjust accordingly.

As with the v2.11 release, we'll provide a public demo instance of the beta once it's released, so keep an eye out for that sometime in the next month or two.

### Roadmap

Further, for the first time in the project's history, we have an [official roadmap](https://github.com/netbox-community/netbox/projects/5)! Roadmap initiatives are categorized into three distinct lanes: core, data model, and integrations. We expect each major release to incorporate work in each category, though focus will naturally shift among areas for each release.

So far, we've identified initiatives for the next three major releases, v3.0 through v3.2. (You can filter the issues display by clicking the v3.x milestone link on an issue.) Please bear in mind that these are subject to change over time, however the roadmap should reasonably convey where the project is headed.

**Comment by FlxPeters** [comment]
This is great. I only have one wish: Please provide more backward compatibility in minor version updates. The breaking changes inside the API on 2.10 and 2.11 makes it really hard to maintain third party integrations with Netbox. With strong typed languages like go it is nearly impossible to provide support for more than one Netbox version at a time. 

Don't get me wrong, progress is great, but please give us some versions with backwards compatibility. Thanks :-)

**Reply by jeremystretch** [reply]
It's a fair point, but the best way to make your concerns known is to actively participate in the development process. We release a beta prior to every major release; this is the ideal time to test and provide feedback, before the final release is published. Without actionable feedback from the developers of API consumers, it's very difficult to identify such work.
